liblastlog2/src/lastlog2.c

```diff
@@ -0,0 +1,567 @@
+  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+  POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <pwd.h>
+#include <errno.h>
+#include <time.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <limits.h>
+#include <sys/stat.h>
+#include <sqlite3.h>
+#include <lastlog.h>
+
+#include "lastlog2.h"
+
+static sqlite3 *
+open_database_ro (const char *path, char **error)
+{
+	sqlite3 *db;
+
+	if (path == NULL)
+		path = LL2_DEFAULT_DATABASE;
+
+	if (sqlite3_open_v2 (path, &db, SQLITE_OPEN_READONLY, NULL) != SQLITE_OK) {
+		if (error)
+			if (asprintf (error, "Cannot open database (%s): %s",
+				      path, sqlite3_errmsg (db)) < 0)
+				fprintf (stderr, "Out of memory\n");
+
+		sqlite3_close (db);
+		return NULL;
+	}
+
+	return db;
+}
+
+static sqlite3 *
+open_database_rw (const char *path, char **error)
+{
+	sqlite3 *db;
+
+	if (path == NULL)
+		path = LL2_DEFAULT_DATABASE;
+
+	if (sqlite3_open (path, &db) != SQLITE_OK) {
+		if (error)
+			if (asprintf (error, "Cannot create/open database (%s): %s",
+				      path, sqlite3_errmsg (db)) < 0)
+				fprintf (stderr, "Out of memory\n");
+
+		sqlite3_close (db);
+		return NULL;
+	}
+
+	return db;
+}
+
+/* Reads one entry from database and returns that.
+   Returns 0 on success, -1 on failure. */
+static int
+read_entry (sqlite3 *db, const char *user,
+	    int64_t *ll_time, char **tty, char **rhost,
+	    char **pam_service, char **error)
+{
+	int retval = 0;
+	sqlite3_stmt *res;
+	static const char *sql = "SELECT * FROM Lastlog2 WHERE Name = ?";
+
+	if (sqlite3_prepare_v2 (db, sql, -1, &res, 0) != SQLITE_OK) {
+		if (error)
+			if (asprintf (error, "Failed to execute statement: %s",
+				      sqlite3_errmsg (db)) < 0)
+				fprintf (stderr, "Out of memory\n");
+
+		return -1;
+	}
+
+	if (sqlite3_bind_text (res, 1, user, -1, SQLITE_STATIC) != SQLITE_OK) {
+		if (error)
+			if (asprintf (error, "Failed to create search query: %s",
+				      sqlite3_errmsg (db)) < 0)
+				fprintf (stderr, "Out of memory\n");
+
+		sqlite3_finalize(res);
+		return -1;
+	}
+
+	int step = sqlite3_step (res);
+
+	if (step == SQLITE_ROW) {
+		const unsigned char *luser = sqlite3_column_text (res, 0);
+		const unsigned char *uc;
+
+		if (strcmp ((const char *)luser, user) != 0) {
+			if (error)
+				if (asprintf (error, "Returned data is for %s, not %s", luser, user) < 0)
+					fprintf (stderr, "Out of memory\n");
+
+			sqlite3_finalize (res);
+			return -1;
+		}
+
+		if (ll_time)
+			*ll_time = sqlite3_column_int64 (res, 1);
+
+		if (tty) {
+			uc = sqlite3_column_text (res, 2);
+			if (uc != NULL && strlen ((const char *)uc) > 0)
+				*tty = strdup ((const char *)uc);
```

Comment:
We need to care about strdup() success. What about
```
*tty = strdup ((const char *)uc);
if (!*tty)
  goto enomem;
...
enomem:
    return -1;  /* o return  -ENOMEM; */
```

You can use it also instead of `fprintf (stderr, "Out of memory\n");` for asprintf.

Reply:
I have added these "jumps". Looks better now :-)